src/reformatters/contrib/noaa/ndvi_cdr/ndvi_cdr/analysis/region_job.py

```diff
@@ -0,0 +1,223 @@
+from collections.abc import Callable, Mapping, Sequence
+from pathlib import Path
+from typing import cast
+
+import fsspec  # type: ignore[import-untyped]
+import numpy as np
+import pandas as pd
+import rasterio  # type: ignore[import-untyped]
+import xarray as xr
+import zarr
+
+from reformatters.common.download import http_download_to_disk
+from reformatters.common.iterating import item
+from reformatters.common.logging import get_logger
+from reformatters.common.region_job import (
+    CoordinateValueOrRange,
+    RegionJob,
+    SourceFileCoord,
+)
+from reformatters.common.types import (
+    AppendDim,
+    Array2D,
+    ArrayFloat32,
+    ArrayInt16,
+    DatetimeLike,
+    Dim,
+    Timestamp,
+)
+
+from . import quality_flags
+from .template_config import (
+    QA_NETCDF_VAR_NAME,
+    NoaaNdviCdrAnalysisTemplateConfig,
+    NoaaNdviCdrDataVar,
+)
+
+QA_DATA_VAR = item(
+    dv for dv in NoaaNdviCdrAnalysisTemplateConfig().data_vars if dv.name == "qa"
+)
+
+
+log = get_logger(__name__)
+
+VIIRS_START_DATE = pd.Timestamp("2014-01-01")
+
+
+class NoaaNdviCdrAnalysisSourceFileCoord(SourceFileCoord):
+    """Coordinates of a single source file to process."""
+
+    time: Timestamp
+    url: str
+
+    def get_url(self) -> str:
+        return self.url
+
+    def out_loc(self) -> Mapping[Dim, CoordinateValueOrRange]:
+        return {"time": self.time}
+
+
+class NoaaNdviCdrAnalysisRegionJob(
+    RegionJob[NoaaNdviCdrDataVar, NoaaNdviCdrAnalysisSourceFileCoord]
+):
+    # We observed deadlocks when using more than 2 threads to read data into shared memory.
+    read_parallelism: int = 1
+
+    def generate_source_file_coords(
+        self,
+        processing_region_ds: xr.Dataset,
+        data_var_group: Sequence[NoaaNdviCdrDataVar],
+    ) -> Sequence[NoaaNdviCdrAnalysisSourceFileCoord]:
+        """Return a sequence of coords, one for each source file required to process the data covered by processing_region_ds.
+
+        The source file names include the processing timestamp, which means that we cannot determine the filename and URL
+        just from a timestamp. When generating SourceFileCoords we need to enumerate the available files and match them to each time we are
+        are attempting to process.
+
+        The source data for this dataset is stored and available from a variety of sources. See the "Data Access" section of
+        https://www.ncei.noaa.gov/products/climate-data-records/normalized-difference-vegetation-index. We are currently using
+        the NOAA S3 bucket.
+        """
+        # set anon to True to not require AWS credentials, as this is a public bucket.
+        fs = fsspec.filesystem("s3", anon=True)
+        public_base_url = "https://noaa-cdr-ndvi-pds.s3.amazonaws.com/data"
+
+        times = processing_region_ds["time"].values
+        years = {pd.Timestamp(t).year for t in times}
```

Comment:
this is like a "next time" thing, but you can get pandas fancy with this
```suggestion
        times = pd.to_datetime(processing_region_ds["time"])
        years = set(times.year)
```

that also lets you skip re-parsing the timestamp with pd.Timestamp down on line 108

Reply:
oh smart, easy enough to do (tried, and tests pass)